src/main/java/com/cloudera/director/openstack/nova/NovaProvider.java

```
@@ -265,19 +383,136 @@ public void allocate(NovaInstanceTemplate template, Collection<String> instanceI
+			}
+			
+			LOG.info("Need to allocate {} volumes for each instances.", volumeNumber);
+			VolumeApi volumeApi = cinderApi.getVolumeApi(region);
+			// We have already confirmed volumeAttApi exists, so get will not fail.
+			VolumeAttachmentApi volumeAttachApi = novaApi.getVolumeAttachmentApi(region).get();
+			final List<String> volumeIds = new ArrayList<String>();
+			
+			//Create all volumes before attaching them to save time.
+			int vol_to_create = volumeNumber * instancesWithPrivateIp.size();
+			LOG.info(">> Start to create {} volumes for the instances.", vol_to_create);
+			for (int i = 0; i < vol_to_create; i++) {
+				CreateVolumeOptions createVolOps = CreateVolumeOptions.Builder
+						.description("SSD")
+						.availabilityZone(azone);
+				Volume currentVolume = volumeApi.create(volumeSize, createVolOps);
+				volumeIds.add(currentVolume.getId());
+			}
+			
+			LOG.info(">> Waiting for {} instance(s) to be attached by volumes.",
+					instancesWithPrivateIp.size());
+			final List<String> activeVolIds = new ArrayList<String>();
+			for (String insId: instancesWithPrivateIp) {				
+				for (int i = 0; i < volumeNumber; i++) {
+					// Get the first volume, and shift the list.
+					String volId = volumeIds.get(0);
+					volumeIds.remove(volId);
+					// We do not set the device so that the devices could be set automatically.
+					String device = "";
+					Volume currentVolume = volumeApi.get(volId);
+					// Wait until Available. The default awaitAvailable wait time is too long (10min).
+					// If not success, we delete it, and regenerate a new volId.
+					boolean createdSuccess = false;
+					if (VolumePredicates.awaitStatus(volumeApi, Volume.Status.AVAILABLE, 60, 5).apply(currentVolume)) {
+						activeVolIds.add(volId);
+						createdSuccess = true;
+					}
+					else{
+						// Retry creating the volume for 5 times.
+						int retry = 5;
+						while (retry > 0 && !createdSuccess) {
```

do not add retry loops - please let exceptions propagate and let the external layer retry the call to allocate()